### PR TITLE
implement common de/serialization of domain objects to vertices

### PR DESCRIPTION
**Pull request by dvoet**

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 1]
```diff
@@ -101,15 +139,15 @@ trait GraphDAO {
...
 
   def entityPipeline(db: Graph, workspaceNamespace: String, workspaceName: String, entityType: String, entityName: String) = {
-    workspacePipeline(db, workspaceNamespace, workspaceName).out(entityType).filter(hasProperty("_name", entityName))
+    workspacePipeline(db, workspaceNamespace, workspaceName).out(entityType).filter(hasPropertyValue("name", entityName))
   }
 
   def methodConfigPipeline(db: Graph, workspaceNamespace: String, workspaceName: String, methodConfigNamespace: String, methodConfigName: String) = {
-    workspacePipeline(db, workspaceNamespace, workspaceName).out(MethodConfigEdgeType).filter(hasProperties(Map("_namespace" -> methodConfigNamespace, "_name" -> methodConfigName)))
+    workspacePipeline(db, workspaceNamespace, workspaceName).out(MethodConfigEdgeType).filter(hasProperties(Map("namespace" -> methodConfigNamespace, "name" -> methodConfigName)))
   }
```
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: Should we define reserved property names (e.g. ```namespace```, ```name```) in a common place somewhere?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 2]
```diff
@@ -122,10 +160,6 @@ trait GraphDAO {
     getSinglePipelineResult(entityPipeline(db, workspaceNamespace, workspaceName, entityType, entityName))
   }
 
-  import scala.reflect.runtime.universe._
-  import scala.reflect.runtime.{universe=>ru}
-  import scala.annotation.meta.field
-
   /**
    * Takes the fields of obj annotated with VertexProperty and sets them on vertex.
    * Vertex property keys are prefixed with an underscore.
```
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: Update doc

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 3]
```diff
@@ -189,22 +295,90 @@ trait GraphDAO {
...
+  private def deserializeMap(vertex: Vertex): Map[String, Attribute] = {
+    vertex.asInstanceOf[OrientVertex].getRecord.setAllowChainedAccess(false)
+
+    val scalarProps: Map[String, Attribute] = getVertexProperties(vertex).map(entry => entry._1 -> AttributeConversions.propertyToAttribute(entry._2))
+
+    val refProps = vertex.getEdges(Direction.OUT).map { edge =>
+      val subV = edge.getVertex(Direction.IN)
+      getVertexClass(subV) match {
+        case VertexSchema.Map =>
+          val attributes: Seq[Attribute] = deserializeSeq(subV)
```
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: Add comment explaining the logic here

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 1 above)
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: +1

Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: in VertexSchema mayhaps

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 4]
```diff
@@ -135,53 +169,125 @@ trait GraphDAO {
...
+        vertex.getVertices(Direction.OUT, key).foreach(removeMapVertex)
       case (key, Some(value: DateTime)) => vertex.setProperty(key, value.toDate)
+      case (key, Some(value: Map[_,_])) => serializeAttributeMap(vertex, key, value.asInstanceOf[Map[String, Attribute]], graph, workspaceName)
+      case (key, Some(value: AttributeReference)) => serializeReference(vertex, key, value, graph, workspaceName)
+      case (key, Some(seq: Seq[_])) => seq.headOption match {
+        case None => // empty, do nothing
+        case Some(x: Workflow) =>
+          serializeDomainObjects(vertex, seq.asInstanceOf[Seq[Workflow]], workflowEdge, (w: Workflow) => w.workflowId, graph, workspaceName)
+          val workspaceVertex = getSinglePipelineResult(workspacePipeline(graph, workspaceName.namespace, workspaceName.name)).get
+          new GremlinPipeline(vertex).out(workflowEdge).linkIn(workflowEdge, workspaceVertex).iterate()
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: i dont know what this is doing - comments?  what happens if we have a seq of objects that don't match these cases, just won't save it or will exceptions be thrown?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 5]
```diff
@@ -135,53 +169,125 @@ trait GraphDAO {
...
+        case Some(x: WorkflowFailure) => serializeDomainObjects(vertex, seq.asInstanceOf[Seq[WorkflowFailure]], workflowFailureEdge, (wf: WorkflowFailure) => (wf.entityType, wf.entityName), graph, workspaceName)
+      }
+      case (key, Some(mrConfig: MethodRepoConfiguration)) => serializeDomainObjects(vertex, Seq(mrConfig), methodRepoConfigEdge, thereShouldOnlyBeOne, graph, workspaceName)
+      case (key, Some(mrMethod: MethodRepoMethod)) => serializeDomainObjects(vertex, Seq(mrMethod), methodRepMethodEdge, thereShouldOnlyBeOne, graph, workspaceName)
+      case (key, Some(ws: WorkspaceName)) => // don't serialize workspace name objects
       case (key, Some(value)) => vertex.setProperty(key, value)
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: if none of the cases above catch, this catches all, what happens if we try to set the property to something we don't expect?  for some circumstances we would get orient exceptions, others maybe not and just unexpected behavior.  Maybe we should be explicit about what we support and throw exception for any type not explicitly defined?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 6]
```diff
@@ -135,53 +169,125 @@ trait GraphDAO {
...
+   * @param workspaceName used to construct domain objects that require a WorkspaceName
+   * @tparam T the type of the object being serializes
+   */
+  private def serializeDomainObjects[T: TypeTag: ClassTag](vertex: Vertex, objs: Seq[T], edgeLabel: String, idFxn: T => Any, graph: Graph, workspaceName: WorkspaceName): Seq[Vertex] = {
+    val existingObjVertexesById = vertex.getVertices(Direction.OUT, edgeLabel).map { objVertex =>
+      val obj = loadFromVertex[T](objVertex, Option(workspaceName))
+      idFxn(obj) -> objVertex
+    } toMap
+
+    (existingObjVertexesById -- objs.map(idFxn)).foreach(_._2.remove())
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: comment would be good here - looks like to me removing the vertex in the map of existing vertexes for any vertex not in the new sub-graph

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 7]
```diff
@@ -135,53 +169,125 @@ trait GraphDAO {
...
+        newVertex
+      })
+
+      saveToVertex(obj, objVertex, graph, workspaceName)
+    }
+  }
+
+  private def serializeAttributeMap(vertex: Vertex, propName: String, map: Map[String, Attribute], graph: Graph, workspaceName: WorkspaceName): Unit = {
+    // remove existing map then repopulate
+    vertex.getVertices(Direction.OUT, propName).headOption.foreach(removeMapVertex)
```
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: Should we check that there isn't more than one outgoing edge labeled ```propName```? Or is it assumed that this will never happen?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 1 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: sure, but that is not part of this story

Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: and the real thing to do is drive it off the case class if we can

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 8]
```diff
@@ -135,53 +169,125 @@ trait GraphDAO {
...
+    vertex.addEdge(propName, mapVertex)
+
+    map.foreach { entry =>
+      entry._2 match {
+        case v: AttributeValue => mapVertex.setProperty(entry._1, AttributeConversions.attributeToProperty(v))
+        case ref: AttributeReference => serializeReference(mapVertex, entry._1, ref, graph, workspaceName)
+        case AttributeValueList(values) => serializeAttributeMap(mapVertex, entry._1, values.zipWithIndex.map(e => e._2.toString -> e._1).toMap, graph, workspaceName)
+        case AttributeReferenceList(references) => serializeAttributeMap(mapVertex, entry._1, references.zipWithIndex.map(e => e._2.toString -> e._1).toMap, graph, workspaceName)
+        case AttributeEmptyList => serializeAttributeMap(mapVertex, entry._1, Map.empty, graph, workspaceName)
+      }
```
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: The ```case (key, value)``` syntax would be nicer than ```entry._1, entry._2```, especially since we have other tuple operations nested inside here.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 9]
```diff
@@ -135,53 +169,125 @@ trait GraphDAO {
...
+    vertex.getVertices(Direction.OUT, propName).headOption.foreach(removeMapVertex)
+
+    val mapVertex = addVertex(graph, VertexSchema.Map)
+    vertex.addEdge(propName, mapVertex)
+
+    map.foreach { entry =>
+      entry._2 match {
+        case v: AttributeValue => mapVertex.setProperty(entry._1, AttributeConversions.attributeToProperty(v))
+        case ref: AttributeReference => serializeReference(mapVertex, entry._1, ref, graph, workspaceName)
+        case AttributeValueList(values) => serializeAttributeMap(mapVertex, entry._1, values.zipWithIndex.map(e => e._2.toString -> e._1).toMap, graph, workspaceName)
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: god i hate tuples

Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: noted

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 10]
```diff
@@ -189,22 +295,90 @@ trait GraphDAO {
...
+
+        case _ => vertexProperty.getOrElse(throw new RawlsException(s"required property [${paramName}] of class [${classT.fullName}] does not have a value in ${vertex}"))
       }
+    }
+
+    try {
+      ctorMirror(parameters: _*).asInstanceOf[T]
+    } catch {
+      case e: Throwable =>
+        println(ctorMirror(parameters: _*))
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: is this just for debugging?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 11]
```diff
@@ -2,32 +2,60 @@ package org.broadinstitute.dsde.rawls.dataaccess
...
+  // container types
+  val Map = vertexClassOf[scala.collection.Map[String,Attribute]]
+
+  def vertexClassOf[T: TypeTag]: String = typeOf[T].typeSymbol.name.decodedName.toString
+}
+
+trait GraphDAO {
   val MethodConfigEdgeType = "_MethodConfig"
 
+  val methodRepMethodEdge: String = "methodRepoMethodEdge"
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: typo

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 5 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: I don't see how to do this. The problem is that we support AnyVal which cannot be used in instance of or pattern matches so it seems like it needs to be the catchall.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 12]
```diff
@@ -189,22 +295,90 @@ trait GraphDAO {
...
     }
 
-    ctorMirror(parameters: _*).asInstanceOf[T]
+    scalarProps ++ refProps
+  }
+
+  private def deserializeSeq(subV: Vertex): Seq[Attribute] = {
+    // Seqs are stored as maps with the index as key
+    // for some reason, orient converts our numeric edge labels to negatives so use Math.abs
+    SortedMap(deserializeMap(subV).map(entry => Math.abs(entry._1.toInt) -> entry._2).toSeq: _*).values.toSeq
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: is this a better solution than for instance "index-0" and stripping off the number from that?

Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: or store a parallel ordered list of rids

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 13]
```diff
@@ -135,53 +169,125 @@ trait GraphDAO {
...
+  private def serializeReference(vertex: Vertex, key: String, ref: AttributeReference, graph: Graph, workspaceName: WorkspaceName): Unit = {
+    val entityVertex = getSinglePipelineResult(entityPipeline(graph, workspaceName.namespace, workspaceName.name, ref.entityType, ref.entityName)).getOrElse {
+      throw new RawlsException(s"${workspaceName.namespace}/${workspaceName.name}/${ref.entityType}/${ref.entityName} does not exist")
+    }
+    vertex.addEdge(key, entityVertex)
   }
 
   /**
    * Constructs and instance of T given a map of properties where the keys correspond to the name of the fields in T
    * (with or without a leading underscore)
```
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: Update doc

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 14]
```diff
@@ -189,22 +295,90 @@ trait GraphDAO {
...
+            case _ => throw new RawlsException(s"org.joda.time.DateTime property [${paramName}] of class [${classT.fullName}] does not have a java.util.Date value in ${vertex}")
+          }
+
+        case mapSymbol if mapSymbol <:< typeOf[Map[String, Attribute]] =>
+          vertex.getVertices(Direction.OUT, paramName).headOption match {
+            case None => Map.empty
+            case Some(mapVertex) => deserializeMap(mapVertex)
+          }
+
+        case entityRef if entityRef =:= typeOf[AttributeReference] =>
```
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: I wonder if ```AttributeReference``` should be renamed to ```EntityReference``` or something, since it is always specific to Entities and it would make a lot of code clearer

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 7 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: If this is the only code that makes them, then it will never be. And that is the point of this code, to do it all centrally, uniformly.

The job of this line is to remove them anyway so after this, none will exist anyway ;)

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 14 above)
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: But I guess that would break the naming scheme. Maybe ```AttributeEntityRef``` ?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 12 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: I don't think it is much different than "index-0" and much less of a PITA than keeping a separate ordered list somewhere.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 15]
```diff
@@ -122,66 +160,146 @@ trait GraphDAO {
...
+        member.name.decodedName.toString.trim -> (fieldMirror.get match {
           case v: Option[Any] => v
           case v => Option(v)
         })
       }
     }
 
-    getProperties(obj, "_").foreach(_ match {
-      case (key, None) => vertex.removeProperty(key)
+    getProperties(obj).foreach(_ match {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: `foreach ( _ match {` can be replaced by `collect {`

File: src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala [hunk 16]
```diff
@@ -172,9 +194,16 @@ class ExpressionParser extends JavaTokenParsers {
...
     // the attribute doesn't exist.
-    FinalResult(lastVertices.map((v: Vertex) => {
-      v.getProperty(attributeName).asInstanceOf[Object]
-    }), s".getProperty($attributeName)")
+    // Note that this requires traversing one additional time, from the "last" vertices to their subordinate attribute map vertices.
+    FinalResult(
+      lastVertices.map((v: Vertex) => {
+        v.getVertices(Direction.OUT, "attributes").headOption match {
+          case Some(mapVertex) => mapVertex.getProperty(attributeName).asInstanceOf[Object]
+          case None => throw new RuntimeException("Boo hoo, what should I do")
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala]: no need for such a formal exception message

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 17]
```diff
@@ -135,53 +169,125 @@ trait GraphDAO {
...
-   * @param otherProperties properties required to construct T but are not in the vertex. Keys must match the
-   *                        names of the fields on T to set
-   * @tparam T
-   * @return an instance of T populated from the vertex and otherProperties
+   * special function passed into idFxn of serializeDomainObjects when there should only be
+   * one object and thus any existing should be overwritten
    */
-  def fromVertex[T: TypeTag](vertex: Vertex, otherProperties: Map[String, Any]): T = {
-    fromPropertyMap(otherProperties ++ getVertexProperties(vertex))
+  private def thereShouldOnlyBeOne(x: Any) = 1
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: typo: `thereCanBeOnlyOne`

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 18]
```diff
@@ -122,66 +160,146 @@ trait GraphDAO {
...
-   *                        names of the fields on T to set
-   * @tparam T
-   * @return an instance of T populated from the vertex and otherProperties
+   * special function passed into idFxn of serializeDomainObjects when there should only be
+   * one object and thus any existing should be overwritten
    */
-  def fromVertex[T: TypeTag](vertex: Vertex, otherProperties: Map[String, Any]): T = {
-    fromPropertyMap(otherProperties ++ getVertexProperties(vertex))
-  }
+  private def thereShouldOnlyBeOne(x: Any) = 1
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: Missed opportunity for a Highlander reference.

highlander-victory

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 10 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: What if this line also throws an exception?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 19]
```diff
@@ -189,22 +307,86 @@ trait GraphDAO {
...
     val parameters = ctor.asMethod.paramLists.head.map { paramSymbol =>
       val paramName = paramSymbol.name.decodedName.toString.trim
-      val prop = vertexProperties.get("_" + paramName).orElse(vertexProperties.get(paramName)) map { value =>
-        if (paramSymbol.typeSignature =:= typeOf[DateTime]) value match {
-          case date: Date => new DateTime(date)
-          case _ => throw new RawlsException(s"org.joda.time.DateTime property [${paramName}] of class [${classT.fullName}] does not have a java.util.Date value in ${vertexProperties}")
+      val vertexProperty: Option[Any] = Option(vertex.getProperty(paramName))
+      paramSymbol.typeSignature match {
+
+        case dateSymbol if dateSymbol =:= typeOf[DateTime] =>
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: I would appreciate a short description explaining what these symbols mean, or at least how `=:=` differs from `<:<`.

Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: http://stackoverflow.com/questions/3427345/what-do-and-mean-in-scala-2-8-and-where-are-they-documented

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphMethodConfigurationDAO.scala [hunk 20]
```diff
@@ -81,52 +43,15 @@ class GraphMethodConfigurationDAO extends MethodConfigurationDAO with GraphDAO {
...
-        configVertex.setProperty("methodConfigVersion", methodConfiguration.methodStoreConfig.methodConfigVersion)
-      })
-    configVertex.setProperty("methodNamespace", methodConfiguration.methodStoreMethod.methodNamespace)
-    configVertex.setProperty("methodName", methodConfiguration.methodStoreMethod.methodName)
-    configVertex.setProperty("methodVersion", methodConfiguration.methodStoreMethod.methodVersion)
-
-    methodConfiguration.inputs.foreach { entry => inputsVertex.setProperty(entry._1, entry._2) }
-    methodConfiguration.outputs.foreach { entry => outputsVertex.setProperty(entry._1, entry._2) }
-    methodConfiguration.prerequisites.foreach { entry => prerequisitesVertex.setProperty(entry._1, entry._2) }
+    val configVertex = getSinglePipelineResult(methodConfigPipe) match {
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphMethodConfigurationDAO.scala]: some places we do getOrElse, some like this, I prefer getOrElse pattern for this

File: src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala (hunk 16 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala]: :+1:

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphMethodConfigurationDAO.scala (hunk 20 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphMethodConfigurationDAO.scala]: :+1:

File: src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala [hunk 21]
```diff
@@ -188,7 +217,7 @@ class ExpressionEvaluator(graph:Graph, parser:ExpressionParser)  {
   def evalFinalAttribute(workspaceNamespace:String, workspaceName:String, rootType:String, rootName:String, expression:String):Try[Seq[AttributeValue]] = {
     parser.parseAttributeExpr(expression)
       .flatMap( runPipe(ExpressionContext(workspaceNamespace, workspaceName, rootType, rootName), _) )
-      .flatMap( ls => Success(ls.map(AttributeConversions.propertyToAttribute)) )
+      .flatMap( ls => Success(ls.map(AttributeConversions.propertyToAttribute(_))) )
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala]: Why this change?

File: src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 22]
```diff
@@ -143,14 +113,15 @@ case class ConflictingEntities(conflicts: Seq[String])
 
 trait Attribute
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: Can make these traits `sealed` to ensure that they are only extended in this file

**Comment by jmthibault79** [comment]
I like the consolidation of `workspaceNamespace / workspaceName` into `WorkspaceName` objects in many places here.

They're still separate in many places, however (especially the DAOs) - I'd like to see a big refactoring sweep to take care of the rest.  Not as part of this PR :)

**Comment by mattsooknah** [comment]
@jmthibault79 that refactoring is currently in progress. stay tuned...

File: src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala (hunk 16 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala]: this is going to bubble up to the NCI one day and we'll end up on the daily wtf

File: src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 23]
```diff
@@ -158,7 +129,7 @@ object AttributeConversions {
     case AttributeBoolean(b) => b
     case AttributeNumber(n) => n.bigDecimal
     case AttributeString(s) => s
-    case AttributeValueList(l) => l.map(attributeToProperty(_)).asJava
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: Where did this go?

Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: Lists are now handled separately in ```GraphDAO.serializeAttributeMap```. It's not needed here because ```AttributeValueList``` is no longer an ```AttributeValue```. 

Although I'm not sure why we moved the conversion code out of ```GraphDAO``` in the first place, since it just makes things confusing.

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: :thumbsup:

File: src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala (hunk 21 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala]: whim... it is equivalent

File: src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 23 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: It got moved to be reused... by expression stuff it looks like, not part of this PR

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 15 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: point taken but I am going to leave it because I think this matches how I think about it better... we are not actually collecting anything, there is no result. We are performing a side effect for each...
